### DISCUSSION TITLE
[KMonad Windows][bug] Unable to Undo or Redo

### DISCUSSION THREAD

**Discussion by ianfromdover** [manna-harbour/miryoku#88] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/62384984/171588294-7b330a93-a5b9-4813-b477-d66469ed44ec.png)

After starting Miryoku using KMonad, it works fine, but crashes and stops working when I try to do the Undo function. I pressed `C +` on my QWERTY keyboard.

**Comment by manna-harbour** [comment, marked as answer]
It looks like that keycode isn't supported by KMonad on windows. But for windows you'll want to build with `MIRYOKU_CLIPBOARD=WIN` to use the standard windows clipboard keyboard shortcuts, which don't include that keycode. See https://github.com/manna-harbour/miryoku/tree/master/docs/reference#clipboard.

**Reply by ianfromdover** [reply]
Thanks for responding to my issue :)

I ran the Build Inputs GitHub Action with `MIRYOKU_CLIPBOARD=WIN` to obtain this `.kbd` file from a clean fork and encountered the error, even though I did not tamper with the file after it was built. Perhaps there may be something wrong with the Action?

Running with `--log-level debug` in PowerShell reveals this message when I execute `redo` and produces a similar message when I execute `undo`.
![image](https://user-images.githubusercontent.com/62384984/172157202-af8fe517-dc2b-43b5-9594-5d8a44689b0f.png)

As an interim solution, I have created my own aliases for undo and redo using `C-z` and `C-y` respectively, but redo does not work and doing undo twice effectively performs a redo. Am I doing something incorrectly? Thank you for your guidance.

**Reply by manna-harbour** [reply]
Looking at your Build Inputs run, the problem is the values have been entered in the wrong format. Just enter `win` in the clipboard field, rather than `MIRYOKU_CLIPBOARD=WIN`. If the value is not recognised you'll just get the default, which does contain `Undo` and `Redo` keycodes. If you want to set defaults in `custom_rules.mk`, or specify options on the `make` command line then you need the long form. It's supposed to be covered in the docs here https://github.com/manna-harbour/miryoku_kmonad#build-inputs-workflow but it's a bit too brief. The Miryoku ZMK workflow docs have been expanded and work in much the same way, so please see those here https://github.com/manna-harbour/miryoku_zmk#miryoku-alternative-layout-and-mapping-options.

Here's a sample build using the options you wanted:
https://github.com/manna-harbour/miryoku_kmonad/actions/runs/2448123070

The `win` option just uses `C-y` and `C-z`, and the behaviour should be the same as if you typed those shortcuts on the keyboard as usual.
